### PR TITLE
Display monthly statistics for the range of months where the project was active

### PR DESCRIPTION
**Pull request by zorun**
Currently, we display a hard-coded "one year" range of monthly statistics
starting from today.  This generally is not the intended behaviour: for
instance, on an archived project, the bills might all be older than one
year, so the table only displays months without any operation.

Instead, display all months between the first and last bills.  There might
be empty months in the middle, but that's intended, because we want all
months to be consecutive.

If there are no bills, simply display an empty table.

### CODE REVIEW

**Comment by zorun** [comment]
Hmm, yes, it could be useful

File: ihatemoney/models.py [hunk 1]
```diff
@@ -257,6 +257,14 @@ def get_member_bills(self, member_id):
             .order_by(Bill.id.desc())
         )
 
+    def get_most_recent_bill(self):
+        """Returns the most recent bill (according to bill date) or None if there are no bills"""
+        return self.get_bills_unordered().order_by(Bill.date.desc()).first()
+
+    def get_least_recent_bill(self):
```
Comment (almet) [inline ihatemoney/models.py:264]: Shouldn't we name this `get_oldest_bill`?

File: ihatemoney/web.py [hunk 2]
```diff
@@ -895,12 +896,23 @@ def strip_ip_addresses():
 @main.route("/<project_id>/statistics")
 def statistics():
     """Compute what each member has paid and spent and display it"""
-    today = datetime.now()
+    if g.project.has_bills():
+        first_date = g.project.get_least_recent_bill().date
+        last_date = g.project.get_most_recent_bill().date
+    else:
+        first_date = datetime.now() - relativedelta(months=12)
+        last_date = datetime.now()
```
Comment (almet) [inline ihatemoney/web.py:904]: I don't quite understand why we need to have last and first date when we don't have bills yet?

File: ihatemoney/web.py [hunk 3]
```diff
@@ -895,12 +896,23 @@ def strip_ip_addresses():
 @main.route("/<project_id>/statistics")
 def statistics():
     """Compute what each member has paid and spent and display it"""
-    today = datetime.now()
+    if g.project.has_bills():
+        first_date = g.project.get_least_recent_bill().date
+        last_date = g.project.get_most_recent_bill().date
```
Comment (almet) [inline ihatemoney/web.py:901]: first and last doesn't really work for me when dealing with dates. I would prefer if we used `oldest_date` and `newest_date` I guess.

File: ihatemoney/web.py [hunk 4]
```diff
@@ -895,12 +896,23 @@ def strip_ip_addresses():
...
+    else:
+        first_date = datetime.now() - relativedelta(months=12)
+        last_date = datetime.now()
+    # Infinite iterator towards the past
+    all_months = (last_date - relativedelta(months=i) for i in itertools.count())
+    # Stop when reaching one month before the first date
+    months = itertools.takewhile(
```
Comment (almet) [inline ihatemoney/web.py:908]: Should we name this `active_months`?

**Comment by almet** [comment]
Hey @zorun, here's a gentle ping on this so we can merge it before it rots (if you have the time, of course!).

**Comment by zorun** [comment]
I don't have time to add tests for this anytime soon. You decide if it can be merged without tests or not :)

File: ihatemoney/web.py (hunk 2 above)
Comment (zorun) [reply ihatemoney/web.py:904]: This is just to mirror the current behaviour (last one year), but yes, we can simply display nothing in this case.

**Comment by almet** [comment]
Rebased.

File: ihatemoney/web.py (hunk 4 above)
Comment (zorun) [reply ihatemoney/web.py]: They are not all active (I explained this in the commit message).

File: ihatemoney/models.py (hunk 1 above)
Comment (zorun) [reply ihatemoney/models.py]: You're right, I like compliated names apparently :)

File: ihatemoney/web.py (hunk 2 above)
Comment (zorun) [reply ihatemoney/web.py]: Changed to display nothing instead.

**Comment by zorun** [comment]
@almet I took into account your remarks, fixed the corner cases, and added a test. A second quick review would be appreciated, thanks!

File: ihatemoney/models.py [hunk 5]
```diff
@@ -262,6 +262,14 @@ def get_member_bills(self, member_id):
             .order_by(Bill.id.desc())
         )
 
+    def get_newest_bill(self):
+        """Returns the most recent bill (according to bill date) or None if there are no bills"""
```
Comment (almet) [inline ihatemoney/models.py:266]: I find it weird to have two return types (None and a list), but that's probably not a problem ? :thinking:

Comment (almet) [reply ihatemoney/models.py:266]: If we always return a list, we might want to get rid of the comments.

File: ihatemoney/tests/budget_test.py [hunk 6]
```diff
@@ -917,6 +917,11 @@ def test_statistics(self):
         # Add a participant with a balance at 0 :
         self.client.post("/raclette/members/add", data={"name": "pépé"})
 
+        # Check that the "monthly expenses" table is empty
+        response = self.client.get("/raclette/statistics")
+        regex = r"<thead>\s*<tr>\s*<th>Period</th>\s*<th>Spent</th>\s*</tr>\s*</thead>\s*<tbody>\s*</tbody>"
```
Comment (almet) [inline ihatemoney/tests/budget_test.py:922]: I'm probably rusted for not reading regexp for some time, but are we sure that a non-empty table will not pass the regex?

File: ihatemoney/tests/budget_test.py [hunk 7]
```diff
@@ -982,6 +987,85 @@ def test_statistics(self):
...
+            ("December 2011", r"\$10.00"),
+            ("November 2011", r"\$0.00"),
+            ("October 2011", r"\$0.00"),
+            ("September 2011", r"\$0.00"),
+            ("August 2011", r"\$50.00"),
+        ]
+        months_fmt = [
+            r"<tr>\s*<td>{}</td>\s*<td>{}</td>\s*</tr>".format(month, amount)
+            for month, amount in months
+        ]
```
Comment (almet) [inline ihatemoney/tests/budget_test.py:1025]: Actually, I find it kinda weird to check the output of the HTML like this, because it means that if we change the format of the table we also need to change the tests, even though we're not changing the actual behavior of the code.

I wonder if we might instead create a model function, and test that it works properly with python-related tests (more unit tests than functional tests, though). We might also want to have this information on the `/api` endpoint.

What do you think?

**Review by almet — COMMENTED** [review]
Hey, here's a quick review. Overall, I think we're good and I'm okay merging it as-is. That being said, if you want to polish things a bit more I've left a few comments. Don't hesitate to ignore them if you feel I'm too picky.

File: ihatemoney/models.py (hunk 5 above)
Comment (zorun) [reply ihatemoney/models.py:266]: It returns a single bill, not a list. Is there a way to clarify this? I mean, the function is called get_something_bill, without S :)

File: ihatemoney/tests/budget_test.py (hunk 6 above)
Comment (zorun) [reply ihatemoney/tests/budget_test.py:922]: Definitely, yes, the \s* only matches whitespace and newlines. So it will only match if there is an empty body.

File: ihatemoney/models.py (hunk 5 above)
Comment (almet) [reply ihatemoney/models.py:266]: Oh my bad, it was very clear but I… I don't know I was somewhere else I guess :sleeping:

File: ihatemoney/models.py [hunk 8]
```diff
@@ -262,6 +262,14 @@ def get_member_bills(self, member_id):
...
+        return self.get_bills_unordered().order_by(Bill.date.desc()).first()
```
Comment (Glandos) [inline ihatemoney/models.py:267]: I am not familiar enough with SQLAlchemy. Could you check that it doesn't fetch a full batch of records from the database? Otherwise, we could manually add a `LIMIT 1` to the query for obvious performance reasons.

Comment (zorun) [reply ihatemoney/models.py:267]: Good idea, I checked and sqlalchemy correctly adds a `LIMIT` itself. I added a comment to make that clear.

**Comment by zorun** [comment]
I have changed the approach a bit:

- move the logic (that computes active months) as a method in the model, because it makes it easier to test
- change the tests to use that new method instead of checking the HTML output
- I am still checking the HTML table once (the empty one) to make sure it's displayed

@almet I guess another review is in order :)

File: ihatemoney/tests/budget_test.py (hunk 7 above)
Comment (zorun) [reply ihatemoney/tests/budget_test.py]: As explained below, I just implement this new testing approach!

For the API, I'm not sure. The client could just request the list of bills and compute any statistics itself.

**Comment by almet** [comment]
It's far better this way, thanks!
